plugin.py

```diff
@@ -0,0 +1,129 @@
+import sublime
+
+from LSP.plugin import AbstractPlugin, register_plugin, unregister_plugin
+from LSP.plugin.core.typing import Any, Optional, Tuple, Union
+
+from shutil import which, copyfileobj
+from urllib.request import urlopen
+from urllib.error import URLError
+import os
+import stat
+import re
+import zipfile
+
+TAG = '1.3.0'
+
+LSP_KOTLIN_BASE_URL = 'https://github.com/fwcd/kotlin-language-server/releases/download/{tag}/server.zip'
+SETTINGS_FILENAME = 'LSP-kotlin.sublime-settings'
+
+LIBRAR_DIR = 'lib'
+DATA_DIR = 'data'
+SESSION_NAME = 'kotlin'
+
+
+RE_VER = re.compile(r'Version\s+((?:\d+\.){2}\d+)')
+
+
+class Kotlin(AbstractPlugin):
+    @classmethod
+    def name(cls):
+        return SESSION_NAME
+
+    @classmethod
+    def basedir(cls) -> str:
+        return os.path.join(cls.storage_path(), __package__)
+
+    @classmethod
+    def server_version(cls) -> str:
+        return TAG
+
+    @classmethod
+    def current_server_version(cls) -> Optional[str]:
+        try:
+            with open(os.path.join(cls.basedir(), 'VERSION'), 'r') as fp:
+                return fp.read()
+        except:
+            return None
+
+    @classmethod
+    def get_kotlin_language_server_binary(cls) -> str:
+        binary = 'kotlin-language-server.bat' if sublime.platform() == 'windows' else 'kotlin-language-server'
+        command = get_setting(
+            'command', [os.path.join(cls.basedir(), 'bin', binary)]
+        )
+        kotlin_binary = command[0].replace('${storage_path}', cls.storage_path())
+        if sublime.platform() == 'windows' and not kotlin_binary.endswith('.bat'):
+            kotlin_binary = kotlin_binary + '.bat'
+        return kotlin_binary
+
+    @classmethod
+    def _is_kotlin_language_server_installed(cls) -> bool:
+        return _is_binary_available(cls.get_kotlin_language_server_binary())
+
+    @classmethod
+    def _is_java_installed(cls) -> bool:
+        return _is_binary_available('java')
+
+    @classmethod
+    def needs_update_or_installation(cls) -> bool:
+        return not cls._is_kotlin_language_server_installed() or (
+            cls.server_version() != cls.current_server_version()
+        )
+
+    @classmethod
+    def install_or_update(cls) -> None:
+        os.makedirs(cls.basedir(), exist_ok=True)
+        download_kotlin_language_server(tag=cls.server_version(), dst=cls.basedir())
+        with open(os.path.join(cls.basedir(), 'VERSION'), 'w') as fp:
+            fp.write(cls.server_version())
+
+        _make_executable(binary=cls.get_kotlin_language_server_binary())
+
+
+def download_kotlin_language_server(tag: Union[str, None], dst: str) -> None:
```

Comment:
It's more natural to use `Optional[str]` instead of an `Union`. But I don't see how `tag` could be `None` so it can be changed to just `str`.